### DISCUSSION TITLE
Implement custom authentication

### DISCUSSION THREAD

**Discussion by idan-h** [OrchardCMS/OrchardCore#13901] [Q&A]
I want to implement a custom module, which will be used in Orchard.Cms.
I want to add custom options in the user settings, and then I want to add a controller with custom dto, from which users will authenticate. I want them to authenticate via openid protocol, and get jwt, and I would like to use the existing infrastructure for this (use it via my controller, because I want to use the custom user data and return it when authenticating, and maybe perform some other actions.) 
Also, I want to implement 2fa via email and sms.
I would like to perform everything from the codebehind or configurations, not via the ui.

I read much of the documentation, watched videos, and I have general idea but I don't know what is the best way to go about it. Would be glad to get the best ideas!
Thanks!

**Comment by hishamco** [comment]
OC already has multiple ways to authenticate users, including authenticating database users and social authentication (Facebook, Twitter, GitHub ... etc)

> Also, I want to implement 2fa via email and SMS.

OC already has `ISmtpService` for sending emails, and a new 2FA module

> I would like to perform everything from the codebehind or configurations, not via the UI.

What you see in the UI is managed by controller and services behind the scene

**Comment by MikeAlhayek** [comment]
As explained in Element. You should use 1.7-previews and utilize OC's 2FA feature.

The 2FA implementation in OC already adds email provider to sends token. You may just need to add a way for the user to enable Email 2FA provider during enabling 2FA "which is do not have in OC yet".

To add SMS, you'll need to complete https://github.com/OrchardCMS/OrchardCore/issues/13710 which should be easy.

Than add a way in the UI for the users to enable 2FA using the phone number.

as far as claims go, you may need to look at the `OpenId` feature in OC and see how you can extend the user info 
https://github.com/OrchardCMS/OrchardCore/blob/7cb1cd2dc69ff5c5305e9feee80ca59d43d793b4/src/OrchardCore.Modules/OrchardCore.OpenId/Controllers/UserInfoController.cs#L19

**Reply by idan-h** [reply]
Just did pull request for that :) #13930

Is there any expectation of when 1.17 will be released?
